src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1188 @@
+    // and values can be stored in the map with only the key stored in the list.
+    //
+    // This implementation currently employs the simple strategy of using both a dictionary and a list, with the
+    // dictionary as the source of truth for the key/value pairs, and the list storing just the keys in order. This
+    // provides O(1) lookups, adding, and indexing, with O(N) insert and removal. Keys are duplicated in memory,
+    // but lookups are optimized to be simple dictionary accesses. Enumeration is O(N), and involves enumerating
+    // the list for order and performing a lookup on each element to get its value. This is the same approach taken
+    // by the non-generic OrderedDictionary and thus keeps algorithmic complexity consistent for someone upgrading
+    // from the non-generic to generic types. It's also important for consumption via the interfaces, in particular
+    // I{ReadOnly}List<T>, where it's common to iterate through a list with an indexer, and if indexing were O(N)
+    // instead of O(1), it would turn such loops into O(N^2) instead of O(N).
+    //
+    // Currently the implementation is optimized for simplicity and correctness, choosing to wrap a Dictionary<>
+    // and a List<> rather than implementing a custom data structure. They could be flattened to partially
+    // deduped in the future if the extra overhead is deemed prohibitive.
+
+    /// <summary>
+    /// Represents a collection of key/value pairs that are accessible by the key or index.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    [DebuggerTypeProxy(typeof(IDictionaryDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class OrderedDictionary<TKey, TValue> :
+        IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>, IDictionary,
+        IList<KeyValuePair<TKey, TValue>>, IReadOnlyList<KeyValuePair<TKey, TValue>>, IList
+        where TKey : notnull
+    {
+        /// <summary>Store for the key/value pairs in the dictionary.</summary>
+        private readonly Dictionary<TKey, TValue> _dictionary;
+        /// <summary>List storing the keys in order.</summary>
+        private readonly List<TKey> _list;
+
+        /// <summary>Lazily-initialized wrapper collection that serves up only the keys, in order.</summary>
+        private KeyCollection? _keys;
+        /// <summary>Lazily-initialized wrapper collection that serves up only the values, in order.</summary>
+        private ValueCollection? _values;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OrderedDictionary{TKey, TValue}"/> class that is empty,
+        /// has the default initial capacity, and uses the default equality comparer for the key type.
+        /// </summary>
+        public OrderedDictionary()
+        {
+            _dictionary = [];
```

Comment:
This constructor will default to whatever initial capacity is being used by each of the underlying collections. Should we make sure that this is set uniformly?

Reply:
We _could_, but maintaining that would be challenging, given that they have different growth strategies (list doubles, dictionary doubles and rounds up to the next prime). Is there a functional concern with keeping them the exact same? If we really want that fine-grained level of control, we can go the other route and open-code the whole implementation rather than relying on a pair of Dictionary/List, I just didn't think it was worth the non-trivial amount of code duplication and complexity right now to do so... do you think otherwise?